source/common/router/router.cc

```diff
@@ -1694,6 +1720,19 @@ void Filter::doRetry() {
     downstream_headers_->setEnvoyAttemptCount(attempt_count_);
   }
 
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.update_expected_rq_timeout_on_retry")) {
+    // If not enabled, then it will re-use the previous headers (if any.)
+
+    Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+    std::chrono::milliseconds elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(
+        dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
```

Comment:
Should this be the downstream started time? We'll potentially start sending upstream requests before the downstream has completed, so this might not even be set?

Reply:
Well spotted.

The timeout timers seem to be started in [onRequestComplete](https://github.com/envoyproxy/envoy/blob/93cdfce6c80fda4107d73f7037744b4953abf92e/source/common/router/router.cc#L876-L885) which is why I used the downstream_request_complete_time_.

So I think, to fit with that behaviour, I should make this conditional on downstream_request_complete_time_ being valid.